cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt

```diff
@@ -228,6 +230,92 @@ fun Node.collectAllPrevFullDFGPaths(): List<List<Node>> {
+ * contains all possible shortest data flow paths (with any
+ * [de.fraunhofer.aisec.cpg.graph.edges.flows.Granularity]) between the end node [this] and the
+ * starting node fulfilling [predicate]. The paths are represented as lists of nodes. Paths which do
+ * not end at such a node are included in [FulfilledAndFailedPaths.failed].
+ *
+ * Hence, if "fulfilled" is a non-empty list, a data flow from [this] to such a node is **possible
+ * but not mandatory**. If the list "failed" is empty, the data flow is mandatory.
+ */
+fun Node.followPrevDFGEdgesUntilHit(
+    collectFailedPaths: Boolean = true,
+    findAllPossiblePaths: Boolean = true,
+    useIndexStack: Boolean = true,
+    predicate: (Node) -> Boolean,
+): FulfilledAndFailedPaths {
+    return followXUntilHit(
+        stack = SimpleStack<IndexedDataflowGranularity>(),
+        x = { currentNode, indexStack ->
+            if (
+                useIndexStack &&
+                    currentNode is InitializerListExpression &&
+                    indexStack?.isEmpty() != true
+            ) {
+                // There's something on the stack. Get the relevant parts
+                currentNode.prevDFGEdges
+                    .filter {
+                        if (
+                            it.start in currentNode.initializers &&
+                                it.granularity is IndexedDataflowGranularity
+                        ) {
+                            // currentNode is the ILE, it is the child and the prevDFG (e.g. read
+                            // from).
+                            // We try to pop from the stack and only select the elements with the
+                            // matching index.
+                            indexStack?.checkAndPop(it.granularity as IndexedDataflowGranularity) ==
+                                true
+                        } else {
+                            true
+                        }
+                    }
+                    .map { it.start }
+            } else {
+                // We don't have to care about poping stuff from the stack. Maybe we can push
+                // something but we accept everything.
+                currentNode.prevDFGEdges.forEach {
+                    if (
+                        it.start is InitializerListExpression &&
+                            it.granularity is IndexedDataflowGranularity
+                    ) {
+                        // CurrentNode is the child and nextDFG goes to ILE => currentNode's written
+                        // to. Push to stack
+                        indexStack?.push(it.granularity as IndexedDataflowGranularity)
+                    }
+                }
+                currentNode.prevDFG
+            }
+        },
+        collectFailedPaths = collectFailedPaths,
+        findAllPossiblePaths = findAllPossiblePaths,
+        predicate = predicate,
+    )
+}
+
+class SimpleStack<T>() {
```

Comment:
Could be private. I am a little bit confused by this is necessary. doesn't ArrayDeque already implement Stack?

Reply:
It does but I wanted to limit the functionality to the minimum. It also doesn't provide a single function for checking and removing the first value on the stack if I remember this right.

Reply:
I would not make the class private in case a user wants to implement an own follow Something function with a stack

Reply:
I'd go with `ArrayDeque` and not implement stuff that's already provided by the language. You could have a `checkAndPop` extension to `ArrayDeque`. I think that's cleaner than re-implementing a stack. But I'm also ok if you want to keep this class.